### ISSUE TITLE
[DQA] adding conflict resolution rules to get-auto-identified-issues command

### ISSUE THREAD

**Issue by writetoritu** [PEDSnet/tools#87]
@bruth Please make suggestions on a suitable format. 

Here is the first set of rules (for code G4-002, i.e. missing data category) to begin with. 
Each rule denotes the changes to be made to the secondary report 

```
For code G4-002 (and a partiuclar field)
x = % of missing data as shown in the log file for the current cycle 
y= % of missing data in the freshly generated template for the current cycle (i.e. y was fed forward from previous cycles) 
acceptable range = x and y are 5% points apart 

Case (i) y% is "under review" (as per the cause field of secondary report)

if (x==y)   : take no action

if ((y==100)  || ((x==100) )
      - create a new issue:  <issue_code=G2-013, issue_description=Unexpected difference from previous ETL run, finding=(x-y)% change in missingness, prevalence=unknown,rank=Medium,status=new>
       - update the existing under review issue: finding=x%

if (x is outside acceptable range of y)
      - create a new issue:  <issue_code=G2-013, issue_description=Unexpected difference from previous ETL run, finding=(x-y)% change in missingness, prevalence=unknown,rank=Medium,status=new>
       -  update the existing under review issue: finding=x%

if (x is within “acceptable" range of y)
       - update the existing under review issue: finding=x%

Case (ii) y% is a “persistent" issue (as per the cause field of secondary report)

if (x==y)  : take no action

if ((y==100) || (x==100) )
      - create a new issue:  <issue_code=G2-013, issue_description=Unexpected difference from previous ETL run, finding=(x-y)% change in missingness, prevalence=unknown,rank=Medium,status=new>
       - create a new issue: <issue_code=G4-002, issue_description=Missing information, finding=x%, prevalence=unknown,rank=Medium,status=new>
       - In the secondary report, delete the y% persistent issue

   if (x is outside acceptable range of y)
         - create a new issue:  <issue_code=G2-013, issue_description=Unexpected difference from previous ETL run, finding=(x-y)% change in missingness, prevalence=unknown,rank=Medium,status=new>
          - create a new issue: <issue_code=G4-002, issue_description=Missing information, finding=x%, prevalence=unknown,rank=Medium,status=new>
          -delete the y% persistent issue

 if (x is within “acceptable" range of y)
       - update the existing persistent issue: finding=x%

Case (iii) Find all G4-002 issues in the secondary file that have no corresponding entries in the log file. For each such issue:
          - delete the y% from secondary file 
          - create a new issue:  <issue_code=G2-013, issue_description=Unexpected difference from previous ETL run, finding=(0-y)% change in missingness, prevalence=unknown,rank=Medium,status=new>

```

**Comment by writetoritu** [comment]
```

# function to resolve conflicts from log file, i.e. compare each issue in log file with matching issues in secondary file
resolve_conflicts_from_log(log_issue):
     if log_issue.check_code==‘G4-002’:
          secondary_issue = get_secondary_issue(log_issue.check_code,log_issue.field_name) # read from the secondary report
          if secondary_issue != null:
               m_prev = secondary_issue.finding # % of missing data from previous cycle
               m_curr = log_issue.finding # % of missing data in current cycle
               diff_check_code= 'code_to_be_created'
               threshold = get_catalog_threshold(diff_check_code,field_name) # returns the range of tolerance for difference for this field

               if secondary_issue.cause==‘under review’:
                     if m_prev==100  or m_curr==100:
                          create_new_issue(issue_code=diff_check_code, finding=(m_curr-m_prev)+'%')
                          secondary_issue.update_issue(finding=m_curr+’%')
                     elif m_curr-m_prev>threshold.upper or m_curr-m_prev < threshold.lower:    # outside acceptable range
                          create_new_issue(issue_code=diff_check_code, finding=(m_curr-m_prev)+'%')
                          secondary_issue.update_issue(finding=m_curr+’%')
                     elif   m_curr-m_prev<threshold.upper or m_curr-m_prev > threshold.lower:
                          secondary_issue.update_issue(finding=m_curr+’%')
                elif secondary_issue.cause==‘persistent:
                     if m_prev==100  or m_curr==100:
                          create_new_issue(issue_code=diff_check_code, finding=(m_curr-m_prev)+'%')
                          create_new_issue(log_issue)
                          secondary_issue.delete_issue()
                     elif m_curr-m_prev>threshold.upper or m_curr-m_prev < threshold.lower:    # outside acceptable range
                          create_new_issue(issue_code=diff_check_code, finding=(m_curr-m_prev)+'%')
                          create_new_issue(log_issue)
                          secondary_issue.delete_issue()
                     elif   m_curr-m_prev<threshold.upper or m_curr-m_prev > threshold.lower:
                          secondary_issue.update_issue(finding=m_curr+’%')
          else: 
                    create_new_issue(log_issue)
# function to resolve conflicts from secondary file, i.e. identify issues having no match in the log file
resolve_conflicts_from_secondary_report(check_code):
          if check_code==‘G4-002':
              no_match_issue_list<-get_no_matching_issues(check_code) # Find all G4-002 issues in the secondary file that have no corresponding entries in the log file.
             diff_check_code= 'code_to_be_created'
             for secondary_issue in  no_match_issue_list :
                     secondary_issue.delete_issue()
                     create_new_issue(issue_code=diff_check_code, finding=(-m_prev)+'%')
```

**Comment by writetoritu** [comment]
@bruth 
For the first version of this tool
- let's develop functions for  "BA-001" check type only that generates the most conflicts
- Also, for conflicts with other check types, we need to print the findings from previous and this cycle so that the user can take a quick action by reading the conflict message.

**Comment by writetoritu** [comment]
@started a draft of the Python code that we discussed the other day 
https://github.com/PEDSnet/Data-Quality/pull/334

**Comment by bruth** [comment]
Implemented in #94
